add codecov.io support (#91)
ros-planning/moveit_ci: README.md

    export ROS_WS=/tmp/ros_ws        # define a new ROS workspace location
    mkdir $ROS_WS                    # and create it
    .moveit_ci/travis.sh

## Enabling codecov.io reporting

> For codecov to work you need to build and link your c++ code with specific parameters.  To enable this we use the ros package [code_coverage](https://github.com/mikeferguson/code_coverage).  To to use the `code-coverage` test in your repo make these two changes:

Reviewer: ```suggestion
For codecov to work you need to build and link your C++ code with specific parameters.  To enable this we use the ROS package [code_coverage](https://github.com/mikeferguson/code_coverage).  To to use the `code-coverage` test in your repo make these two changes:
```

> For codecov to work you need to build and link your C++ code with specific parameters.  To enable this we use the ROS package [code_coverage](https://github.com/mikeferguson/code_coverage).  To to use the `code-coverage` test in your repo make these two changes:

Reviewer 2: ```suggestion
For codecov to work you need to build and link your C++ code with specific parameters.  To enable this we use the ROS package [code_coverage](https://github.com/mikeferguson/code_coverage).  Using the `code-coverage` test in your repo requires the following two changes:
```

1. Add `<test_depend>code_coverage</test_depend>` to your package.xml

> 1. Add this to your `CMakeLists.txt`:

Reviewer 2: ```suggestion
2. Add this to your `CMakeLists.txt`:
```

```cmake
# to run: catkin_make -DENABLE_COVERAGE_TESTING=ON package_name_coverage

> if(CATKIN_ENABLE_TESTING AND ENABLE_COVERAGE_TESTING)

Reviewer 2: Maybe split up this if condition into two separate ones to match your code snippet below?

Author: This is the exact code that is in the moveit codebase already.  I just copied it from there.  What I did in the unit test is the minimal amount needed.  What is kind of nice about this if statement (like what is in the moveit codebase) is it adds the codecoverage to one section with out being two blocks of code like what's in the example from code_coverage ros package's README.md.

Author: This is coppied code from the cmake files in moveit.  I like how it keeps the code coverage section separate from the testing if section.  I can change it if you'd perfer.  I made the unit test match this.

  find_package(code_coverage REQUIRED)   # catkin package ros-*-code-coverage
  include(CodeCoverage)
  APPEND_COVERAGE_COMPILER_FLAGS()
  set(COVERAGE_EXCLUDES "*/test/*")

>   add_code_coverage(NAME ${PROJECT_NAME}_coverage)

Reviewer 2: Is this really needed? You didn't use this command below for the unit test.

Author: Yes, I'll add it to the unit test.

>   add_code_coverage(NAME ${PROJECT_NAME}_coverage)

Reviewer 2: This line is not needed.

Author: removed

    export ROS_WS=/tmp/ros_ws        # define a new ROS workspace location
    mkdir $ROS_WS                    # and create it
    .moveit_ci/travis.sh

## Enabling codecov.io reporting

For codecov to work you need to build and link your C++ code with specific parameters.  To enable this we use the ROS package [code_coverage](https://github.com/mikeferguson/code_coverage).  Using the `code-coverage` test in your repo requires the following three changes:

1. Add `<test_depend>code_coverage</test_depend>` to your package.xml
2. Add this to your `CMakeLists.txt`:

```cmake
# to run: catkin_make -DENABLE_COVERAGE_TESTING=ON package_name_coverage
if(CATKIN_ENABLE_TESTING AND ENABLE_COVERAGE_TESTING)
  find_package(code_coverage REQUIRED)   # catkin package ros-*-code-coverage

>   include(CodeCoverage)

Reviewer 3: I just released 0.3.0 with @rhaschke improvements - once that propagates through the build farm and gets sync'd you'll be able to drop this line

Author: Which line?  The find_package or the include?  Can you link me to your changes?

Reviewer 3: The include -- see https://github.com/mikeferguson/code_coverage/pull/15 for details (again, I just did the release, it's probably 2 weeks before it syncs all the way through the build farm).

Author: Ah, cool.  Hopefully I'll get this merged quickly and then I can make a small PR to remove this line once your change propagates.

    export ROS_WS=/tmp/ros_ws        # define a new ROS workspace location
    mkdir $ROS_WS                    # and create it
    .moveit_ci/travis.sh

## Enabling codecov.io reporting

> To enable codecov enabling you'll need to add this to your CMakeLists.txt file:

Reviewer 4: "enable codecov enabling"?

    option(CMAKE_CODE_COVERAGE_CONFIG
      "Generate code coverage reporting for codecov.io"
      OFF
    )

    # Code coverage build settings
    add_library(coverage_config INTERFACE)
    if(CMAKE_CODE_COVERAGE_CONFIG)
      # Add required flags (GCC & LLVM/Clang)
      target_compile_options(coverage_config INTERFACE
        -O0        # no optimization
        -g         # generate debug info
        --coverage # sets all required flags
      )
      target_link_libraries(coverage_config INTERFACE --coverage)
    endif()

> Then you'll need to change the `target_link_libraries` command to turn on code coverage specifically for your project without propigating these settings to external dependencies:

Reviewer 4: ```suggestion
Then you'll need to change the `target_link_libraries` command to turn on code coverage specifically for your project without propagating these settings to external dependencies:
```

    target_link_libraries(
      ${PROJECT_NAME}
      PUBLIC
      ${catkin_LIBRARIES}
      ${Boost_LIBRARIES}
    )

    target_link_libraries(

>       ${PROJECT_NAME}

Author: It builds them as part of the PUBLIC link policy above.  This was to separate what gets built with the --coverage flag.  I'm going to do some more testing and see if I can simplify this.

  include(CodeCoverage)
  APPEND_COVERAGE_COMPILER_FLAGS()

>   set(COVERAGE_EXCLUDES "*/test/*")

Reviewer 2: I'm pretty sure that the `COVERAGE_EXCLUDES` variable has no effect here. This is only relevant for use with the coverage report target created by `add_code_coverage()`. But this we don't use here, because we use the bash script provided by codecov.io.
Did you validated the effect of `COVERAGE_EXCLUDES`?

Author: You are right :(.  If you look at the files that were "tested" in the above linked codecov.io report you will find tests.  Good catch.  I think we'll need to bring back the codecov yaml file.  I'll test that now and update this.
